src/OneDriveSdk/Helpers/UploadSessionHelper.cs

```diff
@@ -0,0 +1,257 @@
+﻿// ------------------------------------------------------------------------------
+//  Copyright (c) Microsoft Corporation.  All Rights Reserved.  Licensed under the MIT License.  See License in the project root for license information.
+// ------------------------------------------------------------------------------
+
+namespace Microsoft.OneDrive.Sdk.Helpers
+{
+    using Microsoft.Graph;
+
+    using System;
+    using System.Collections.Generic;
+    using System.IO;
+    using System.Linq;
+    using System.Threading;
+    using System.Threading.Tasks;
+
+    public class UploadSessionHelper
+    {
+        private const int DefaultMaxChunkSize = 10 * 1024 * 1024;
+
+        public UploadSession Session { get; private set; }
+        private IBaseClient client;
+        private Stream uploadStream;
+        private int totalUploadLength;
+        private readonly int maxChunkSize;
+        private List<Tuple<int, int>> rangesRemaining;
+        
+        /// <summary>
+        /// Helps with resumable uploads. Generates chunk requests based on <paramref name="session"/>
+        /// information, and can control uploading of requests using <paramref name="client"/>
+        /// </summary>
+        /// <param name="session">Session information.</param>
+        /// <param name="client">Client used to upload chunks.</param>
+        /// <param name="uploadStream">Readable, seekable stream to be uploaded.</param>
+        /// <param name="totalUploadLength">Total length of file to uploaded. <paramref name="uploadStream"/> must have at least
+        /// this many bytes.</param>
+        /// <param name="maxChunkSize">Max size of each chunk to be uploaded. If less than 0, default value of 10 MB is used. Increment of 320 kb is
+        /// recommended.</param>
+        public UploadSessionHelper(UploadSession session, IBaseClient client, Stream uploadStream, int totalUploadLength, int maxChunkSize = -1)
```

Comment:
What if the stream is longer than the request? Like, perhaps I initialize a memory stream with 100MB but only use 80MB of it. Wouldn't it be better to just specify the size instead of forcing them to resize the stream?